src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -228,6 +235,19 @@ protected internal int GetInvocationCounter()
             return counter;
         }
 
+        /// <summary>
+        /// The implementation of System.Runtime.GetRandom.
+        /// Gets the random number genrated form the VRF
+        /// Primary geterates this random number with `prevHash`, the hash of the previous (validators.Length/3 + 1) block
+        /// </summary>
+        /// <returns>The last four bytes of the random number.</returns>
+        protected internal uint GetRandom()
+        {
+            var nonce = next_nonce;
+            next_nonce = Cryptography.Helper.Sha256(nonce);
```

Comment:
Random=SHA256(nonce).

Problems:
- The next execution, in the same context (this is important), it will be predictable. 
- An execution in different context, it will return the same result.


Attack:
- Send a tx and call random, store the result in the storage.
- Send a second tx and read from the storage the result of the next random call.
- If both tx enter in the same block, you will win the lotery.

Reply:
Are you saying that `next_nonce = BitConverter.GetBytes(persistingBlock.Nonce);` will be called multiple times ?

Reply:
@shargon corrected and tested.

Reply:
Agreed @shargon. But I believe that a lottery should also use other calculus,such as  tx hash merged with the nonce, getting its Sha again.

Reply:
Agreed @shargon, but it looks like that the Lotery Contract should block storage read from other contracts, right? Then, there would not be way to read the storage and share this information.

Reply:
Will a node that starts synchronizing the chain and reaches the current block have the same `nonce` as one that goes in real time? still, what prevents you from calling random and computing the next value that is called?

I think that the only way to produce a **secure random** require `submarine sends` (https://www.youtube.com/watch?v=N8PDKoptmPs)

Reply:
> Will a node that starts synchronizing the chain and reaches the current block have the same `nonce` as one that goes in real time? still, what prevents you from calling random and computing the next value that is called?
> 
> 
> 
> I think that the only way to produce a **secure random** require `submarine sends` (https://www.youtube.com/watch?v=N8PDKoptmPs) 


I don't understand, what a node could do even if it knows the real-time nonce? The nonce only works for the current or past block, even if a node learns the nonce, it would be already too late to do anything, and the nonce is calculated with VRF, of which the primary can not decide the result. As the issue that the primary might change the transactions order or other type of dishonest things, I think it's another security issue.

Reply:
@Liaojinghui, about this too late I am not sure.

If the block is finalized, yes.
However, we need to make sure that nonce is just disclosed after 2f+1 commits.

Otherwise, nonce could be exposed before finalized and a change view could incur in the Backup adding a biased tx.

On the other hand, I was planning to handle that on the neo-modules PR.

Reply:
I just watched the submarine, @shargon, it looks like that, as in anonymous pools, observers can still imagine the size of the change, although they can not differentiate commits or simple sends of ether.
That would require huge use to become something really safe.

Reply:
> @Liaojinghui, about this too late I am not completely sure.
> 
> If the block is finalized, yes.
> However, we need to make sure that nonce is just disclosed after 2f+1 commits.
> 
> Otherwise, nonce could be exposed before finalized and a change view could incur in the Backup adding a biased tx.
> 
> On the other hand, I was planning to handle that on the neo-modules PR.

If a changing view occurs, a new nonce will be calculated, the different primary will have different calculation results even with the same input.

Reply:
In this sense, I did not understand very well the attack @shargon is mentioning.